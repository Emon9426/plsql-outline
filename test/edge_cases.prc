/* 边界情况测试 */
CREATE OR REPLACE PROCEDURE edge_case_procedure AS
  /* 注释中的关键字：BEGIN END FUNCTION PROCEDURE DECLARE */
  
  -- 情况1：只有END
  PROCEDURE simple_end IS
  BEGIN
    NULL;
  END; -- 只有END
  
  -- 情况2：END后跟函数名
  FUNCTION func_with_named_end RETURN NUMBER IS
  BEGIN
    RETURN 1;
  END func_with_named_end; -- END后跟函数名
  
  -- 情况3：多层嵌套与混合END
  PROCEDURE mixed_end IS
    FUNCTION inner_func RETURN NUMBER IS
    BEGIN
      RETURN 2;
    END inner_func; -- END后跟函数名
    
    PROCEDURE inner_proc IS
    BEGIN
      NULL;
    END; -- 只有END
  BEGIN
    DECLARE
      v_temp NUMBER := 0;
    BEGIN
      v_temp := inner_func();
      DBMS_OUTPUT.PUT_LINE('值: ' || v_temp);
    END; -- 只有END
  END mixed_end; -- END后跟过程名
  
  -- 情况4：注释中的嵌套结构
  /*
    FUNCTION commented_function IS
    BEGIN
      -- 这只是一个注释
      NULL;
    END commented_function;
  */
  
BEGIN
  -- 主执行块
  DECLARE
    v_result NUMBER;
  BEGIN
    v_result := func_with_named_end();
    DBMS_OUTPUT.PUT_LINE('函数结果: ' || v_result);
    mixed_end();
    
    -- 异常块测试
    BEGIN
      RAISE NO_DATA_FOUND;
    EXCEPTION
      WHEN NO_DATA_FOUND THEN
        DBMS_OUTPUT.PUT_LINE('捕获到NO_DATA_FOUND异常');
    END; -- 只有END
  END; -- 只有END
  
  /* 结束注释 */
EXCEPTION
  WHEN OTHERS THEN
    DBMS_OUTPUT.PUT_LINE('主异常块: ' || SQLERRM);
END edge_case_procedure; -- END后跟过程名
/
