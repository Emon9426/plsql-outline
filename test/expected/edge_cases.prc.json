{
  "nodes": [
    {
      "label": "edge_case_procedure",
      "type": "procedure",
      "icon": "symbol-method",
      "children": [
        {
          "label": "simple_end",
          "type": "procedure",
          "icon": "symbol-method",
          "children": [
            {
              "label": "BEGIN",
              "type": "begin",
              "icon": "symbol-namespace",
              "children": [],
              "range": {
                "startLine": 6,
                "endLine": 8
              }
            }
          ],
          "range": {
            "startLine": 5,
            "endLine": 8
          }
        },
        {
          "label": "func_with_named_end",
          "type": "function",
          "icon": "symbol-function",
          "children": [
            {
              "label": "BEGIN",
              "type": "begin",
              "icon": "symbol-namespace",
              "children": [],
              "range": {
                "startLine": 12,
                "endLine": 12
              }
            }
          ],
          "range": {
            "startLine": 11,
            "endLine": 14
          }
        },
        {
          "label": "mixed_end",
          "type": "procedure",
          "icon": "symbol-method",
          "children": [
            {
              "label": "inner_func",
              "type": "function",
              "icon": "symbol-function",
              "children": [
                {
                  "label": "BEGIN",
                  "type": "begin",
                  "icon": "symbol-namespace",
                  "children": [],
                  "range": {
                    "startLine": 19,
                    "endLine": 19
                  }
                }
              ],
              "range": {
                "startLine": 18,
                "endLine": 21
              }
            },
            {
              "label": "inner_proc",
              "type": "procedure",
              "icon": "symbol-method",
              "children": [
                {
                  "label": "BEGIN",
                  "type": "begin",
                  "icon": "symbol-namespace",
                  "children": [],
                  "range": {
                    "startLine": 24,
                    "endLine": 26
                  }
                }
              ],
              "range": {
                "startLine": 23,
                "endLine": 26
              }
            },
            {
              "label": "BEGIN",
              "type": "begin",
              "icon": "symbol-namespace",
              "children": [
                {
                  "label": "DECLARE",
                  "type": "declare",
                  "icon": "symbol-variable",
                  "children": [
                    {
                      "label": "BEGIN",
                      "type": "begin",
                      "icon": "symbol-namespace",
                      "children": [],
                      "range": {
                        "startLine": 30,
                        "endLine": 33
                      }
                    }
                  ],
                  "range": {
                    "startLine": 28,
                    "endLine": 28
                  }
                }
              ],
              "range": {
                "startLine": 27,
                "endLine": 27
              }
            }
          ],
          "range": {
            "startLine": 17,
            "endLine": 34
          }
        },
        {
          "label": "BEGIN",
          "type": "begin",
          "icon": "symbol-namespace",
          "children": [
            {
              "label": "DECLARE",
              "type": "declare",
              "icon": "symbol-variable",
              "children": [
                {
                  "label": "BEGIN",
                  "type": "begin",
                  "icon": "symbol-namespace",
                  "children": [
                    {
                      "label": "BEGIN",
                      "type": "begin",
                      "icon": "symbol-namespace",
                      "children": [],
                      "range": {
                        "startLine": 55,
                        "endLine": 61
                      }
                    },
                    {
                      "label": "EXCEPTION",
                      "type": "exception",
                      "icon": "error",
                      "children": [],
                      "range": {
                        "startLine": 64,
                        "endLine": 64
                      }
                    }
                  ],
                  "range": {
                    "startLine": 49,
                    "endLine": 49
                  }
                }
              ],
              "range": {
                "startLine": 47,
                "endLine": 47
              }
            }
          ],
          "range": {
            "startLine": 45,
            "endLine": 45
          }
        },
        {
          "label": "EXCEPTION",
          "type": "exception",
          "icon": "error",
          "children": [],
          "range": {
            "startLine": 57,
            "endLine": 60
          }
        }
      ],
      "range": {
        "startLine": 1,
        "endLine": 60
      }
    }
  ],
  "errors": []
}